### ISSUE TITLE
How is it possible to send a blob?

### ISSUE THREAD

**Issue by seriouz** [aspnet/SignalR#2519]
Hello, is it currently possible to send a blob over the websocket connection? (https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)
The blob is filled with chunks from a webcam stream encoded in webm/v9.

**Comment by anurse** [comment]
Do you just want to send a single blob as a `byte[]` in a single invocation? I believe we don't support that now, only ArrayBuffer is supported. What error are you seeing when you try?

**Comment by seriouz** [comment]
ArrayBuffer does not work either. The data object which gets transmitted is `{}`.

**Comment by anurse** [comment]
The message you're sending is 648KB in length. That's a very large message. SignalR is not designed to handle large single messages like that, it's designed for more frequent smaller messages. You'll have to break the message up into smaller chunks and send each chunk in a separate invocation (using some ID number and sequence numbers to keep track of which chunks are a part of which invocations and what order they come in).

If you're trying to send a video stream over SignalR, you're going to have to do a lot of manual work. It's simply not something SignalR is currently designed for. We've talked about ways to improve that, but the overhead is going to be quite large, compared to using something like WebRTC which is designed for exactly this purpose.

**Comment by Chemsorly** [comment]
It is possible, but not recommended. You can bump the message size limit by increasing ApplicationMaxBufferSize (see [#2266](https://github.com/aspnet/SignalR/issues/2266#issuecomment-388889329) for details).

**Comment by popcatalin81** [comment]
To add to this conversation, I've also recently used SignalR to send ffmpeg compressed video frame packets with VP9 using SignalR and the solution was indeed to increase the ApplicationMessage size slightly, then send each packet as byte[] array.

A byte-streaming mode from client to server would be very welcome for such a scenario.

**Comment by anurse** [comment]
We've been looking at streaming from client-to-server. We're not yet sure how exactly it will land, but we've been working on _prototype_ support for it. No guarantees it will land in 2.2, but it's something we're playing around with. It's definitely something you can build yourself though, since you can simply call separate Hub methods to initialize a stream and send smaller chunks of data.

**Comment by popcatalin81** [comment]
@anurse That's what I did. However, the problem is that my wrapper is allocating a `byte[]` array for each message, I would preffer0 a non-allocating alternative and currently, SignalR does not expose extensibility points to do this (Unless I've missed something)

**Comment by anurse** [comment]
> However, the problem is that my wrapper is allocating a byte[] array for each message

What do you mean by this? Where is the `byte[]` allocation that you want to remove? On the client or the server?

**Comment by popcatalin81** [comment]
@anurse, I would like to eliminate the byte[] array allocation In both places ideally client code like this:

```CSharp
	var frame = new Frame
	{
		Data = new byte[length]
	}
	CopyUnmanagedPointerToBtyeArray(src, length, frame.Data);
	await SendPacket(frame);
```

Would be replaced with a non allocating version byte stream version. Same on the Server.

**Comment by anurse** [comment]
I see, so you have data in an unmanaged buffer you want to send without having to copy it to a managed buffer? This isn't really feasible right now. You could wrap it up in `ReadOnlyMemory<byte>` but I don't know for sure if that's possible or if JSON.NET/MessagePack-CSharp (our serializer libraries) will handle it properly.

If the only reason the buffer is unmanaged is because you need to call a native API to fill it, you could also try using a managed `byte[]` and pin it in order to get a native pointer you can pass into the native function to fill it up. Then you actually have a managed buffer, but the native code can fill it up:

```csharp
var buffer = new byte[length];
fixed(byte* ptr = buffer)
{
    NativeAPIToFillTheBuffer(ptr, length);
}
```

Once you've filled the buffer, you can pass the `byte[]` array around without allocating new things. You can also pool the buffers (as long as you ensure you don't return the buffer to the pool until **after** `await`ing the `InvokeAsync`/`SendAsync` call).

**Comment by anurse** [comment]
On the server, there's no way to pool the `byte[]` allocations we use when deserializing the message so that's not something that can be avoided at the moment.

**Comment by popcatalin81** [comment]
@anurse there are a few shortcomings to using a byte array pool in my solution.

ffmpeg API does not allow you to pass a pointer to a buffer to use by compression API, it will always allocate a packet memory itself and refcount the packet, by design.

Using a buffer pool of byte[] arrays means they will have fixed sizes, compressed packets are not fixed sizes they can vary up to two orders of magnitude in size from around 1KB up to 200 - 300 KB with chosen settings, averaging at around 30 kb. I have chosen to send a packet in a single call and not bother with chunking at this stage (there was no observable benefit for my solution as the server is far from being maxed out). This means using a buffer pool, either buffer are too big and extra data is serialized wastefully, or they are too small and to many messages are sent, again wastefully.

In conclusion, the solution to increase application buffer size worked best for my use case.

**Comment by aspnet-hello** [comment]
We periodically close 'discussion' issues that have not been updated in a long period of time.

We apologize if this causes any inconvenience. We ask that if you are still encountering an issue, please log a new issue with updated information and we will investigate.
